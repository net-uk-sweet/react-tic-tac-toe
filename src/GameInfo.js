import React from 'react';

const GameInfo = props => {

    const renderMove = (step, index, list, ascendingOrder) => {

        const totalMoves = list.size - 1;
        const move = step.get('squares').get(step.get('index'));
        const isStartMove = ascendingOrder ? index === 0 : index === totalMoves;
        const mutatedIndex = ascendingOrder ? index : totalMoves - index;
  
        const details = isStartMove ? 'Go to start' : 
          `Go to move ${mutatedIndex} - column - ${move.get('column')} - row - ${move.get('row')}`;
      
        return (
          <li key={index}  
            className={mutatedIndex === props.currentMove ? 'selected qa-selected' : ''}>
              <button onClick={() => props.onMoveClick(mutatedIndex)}>{details}</button>
          </li>
        );
      }

    const moves = (props.ascendingOrder ? props.history : props.history.reverse()).map((step, move, list) => {
        return renderMove(step, move, list, props.ascendingOrder);
    });

    return (
        <React.Fragment>
            <div>{props.status}</div>
            <ol>{moves}</ol>
        </React.Fragment>
    );
}

export default GameInfo;