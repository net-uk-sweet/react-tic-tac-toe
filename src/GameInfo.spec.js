import React from 'react';
import Immutable  from 'immutable';
import { mount, shallow } from 'enzyme';

import GameInfo from './GameInfo';

let shared;
let mockedClick;

beforeEach(() => {
    mockedClick = jest.fn();
    shared = {
        startLabel: 'Go to start',
        props: {
            history: Immutable.fromJS([
                {
                    "squares": [
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    }
                    ]
                },
                {
                    "index": 0,
                    "squares": [
                    {
                        "token": "x",
                        "row": 1,
                        "column": 1
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    }
                    ]
                },
                {
                    "index": 4,
                    "squares": [
                    {
                        "token": "x",
                        "row": 1,
                        "column": 1
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "o",
                        "row": 2,
                        "column": 2
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    }
                    ]
                },
                {
                    "index": 2,
                    "squares": [
                    {
                        "token": "x",
                        "row": 1,
                        "column": 1
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "x",
                        "row": 1,
                        "column": 3
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "o",
                        "row": 2,
                        "column": 2
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    }
                    ]
                },
                {
                    "index": 5,
                    "squares": [
                    {
                        "token": "x",
                        "row": 1,
                        "column": 1
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "x",
                        "row": 1,
                        "column": 3
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "o",
                        "row": 2,
                        "column": 2
                    },
                    {
                        "token": "o",
                        "row": 2,
                        "column": 3
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    }
                    ]
                },
                {
                    "index": 8,
                    "squares": [
                    {
                        "token": "x",
                        "row": 1,
                        "column": 1
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "x",
                        "row": 1,
                        "column": 3
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "o",
                        "row": 2,
                        "column": 2
                    },
                    {
                        "token": "o",
                        "row": 2,
                        "column": 3
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "x",
                        "row": 3,
                        "column": 3
                    }
                    ]
                },
                {
                    "index": 7,
                    "squares": [
                    {
                        "token": "x",
                        "row": 1,
                        "column": 1
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "x",
                        "row": 1,
                        "column": 3
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "o",
                        "row": 2,
                        "column": 2
                    },
                    {
                        "token": "o",
                        "row": 2,
                        "column": 3
                    },
                    {
                        "token": null,
                        "column": null,
                        "row": null
                    },
                    {
                        "token": "o",
                        "row": 3,
                        "column": 2
                    },
                    {
                        "token": "x",
                        "row": 3,
                        "column": 3
                    }
                    ]
                }
            ]),
            currentMove: 2,
            ascendingOrder: true,
            onMoveClick: mockedClick
        }
    }
    shared.wrapper = shallow(<GameInfo {...shared.props} />);
});

it('matches snapshot', () => {
  expect(shared.wrapper.getElement()).toMatchSnapshot();
});

it('highlights the current move', () => {
    const currentMove = shared.wrapper.find('li').at(shared.props.currentMove);
    expect(currentMove.hasClass('qa-selected')).toBeTruthy();
});

it('labels the first move correctly', () => {
    expect(shared.wrapper.find('li').at(0).text()).toEqual(shared.startLabel);
})

it('calls the click handler when a button is selected', () => {
   shared.wrapper.find('button').at(2).simulate('click');
   expect(mockedClick).toHaveBeenCalledWith(2); 
});

describe('when the move order is reversed', () => {
    beforeEach(() => {
        const props = Object.assign(shared.props, { ascendingOrder: false });
        shared.historySize = shared.props.history.size;
        shared.wrapper = shallow(<GameInfo {...props} />);
    });

    it('matches snapshot', () => {
        expect(shared.wrapper.getElement()).toMatchSnapshot();
    });

    it('highlights the current move', () => {
        const index = shared.historySize - 1 - shared.props.currentMove;
        const currentMove = shared.wrapper.find('li').at(index);
        expect(currentMove.hasClass('qa-selected')).toBeTruthy();
    });

    it('labels the first move correctly', () => {
        expect(shared.wrapper.find('li').at(shared.historySize - 1).text()).toEqual(shared.startLabel);
    });
});
