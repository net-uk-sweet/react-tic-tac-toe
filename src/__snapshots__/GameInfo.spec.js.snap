// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Unknown>
  <div />
  <ol>
    Immutable.List [
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to start
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 1 - column - 1 - row - 1
        </button>
      </li>,
      <li
        className="selected qa-selected"
      >
        <button
          onClick={[Function]}
        >
          Go to move 2 - column - 2 - row - 2
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 3 - column - 3 - row - 1
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 4 - column - 3 - row - 2
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 5 - column - 3 - row - 3
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 6 - column - 2 - row - 3
        </button>
      </li>,
    ]
  </ol>
</Unknown>
`;

exports[`when the move order is reversed matches snapshot 1`] = `
<Unknown>
  <div />
  <ol>
    Immutable.List [
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 6 - column - 2 - row - 3
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 5 - column - 3 - row - 3
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 4 - column - 3 - row - 2
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 3 - column - 3 - row - 1
        </button>
      </li>,
      <li
        className="selected qa-selected"
      >
        <button
          onClick={[Function]}
        >
          Go to move 2 - column - 2 - row - 2
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to move 1 - column - 1 - row - 1
        </button>
      </li>,
      <li
        className=""
      >
        <button
          onClick={[Function]}
        >
          Go to start
        </button>
      </li>,
    ]
  </ol>
</Unknown>
`;
